roofit/roofitcore/src/RooProfileLL.cxx

```diff
@@ -221,15 +184,13 @@ Double_t RooProfileLL::evaluate() const
   _neval = _minimizer->evalCounter() ;
 
   // Restore original values and constant status of observables
-  TIterator* iter = obsSetOrig->createIterator() ;
-  RooRealVar* var ;
-  while((var=dynamic_cast<RooRealVar*>(iter->Next()) ) )  {
-    RooRealVar* target = (RooRealVar*) _obs.find(var->GetName()) ;
+  for(auto const& arg : obsSetOrig) {
+    auto var = dynamic_cast<RooRealVar*>(arg);
+    if(var == nullptr) break;
```

Comment:
This pattern was used in a lot of old loops, and although it looks like a double check (both for checking that the loop isn't finished and that the type is correct), it's usually only checking if the loop is finished. In almost all cases, you can `static_cast` and run through the full collection. For good measure, I put an `assert` with a `dynamic_cast` into the loop. It brings extra debug info, and it doesn't cost anything in release builds.

Note also that if they were really checking the types, it wouldn't make any sense if you put `{var, var, otherType, var}` in a collection, because the last var would never be reached. That's most probably a bug.

Long story short: Those loops usually work with range-base and `static_cast`, and they will be much faster this way. Consider this as a standard way to modernise those kinds of loops (and make the `var` `const` whenever you can; here you can't). Also note that although `auto const&` looks like an awesome thing for "normal" collections, it's actually obfuscating the fact that you have a collection of pointers.
```suggestion
  for(auto const arg : obsSetOrig) {
    auto var = static_cast<RooRealVar*>(arg);
    assert(dynamic_cast<RooRealVar*>(arg));
```

Reply:
And if you want (I never went there), you can write an adapter that exposes a special kind of iterator:
```c++
for (auto const pvar : CastCollection<const RooRealVar*>(obsSetOrig) ) {
  pvar->GetName();
}
  ```
  Maybe you can get some inspiration from #8421

Reply:
Thanks for the comment! Very good point, I was too cautions here and didn't want to change the logic of the code. I'll do it now.